### PR TITLE
Warm up instance cache on master restart

### PR DESCRIPTION
**Pull request by DrDet**
Here is the implementation of more smooth graceful restart to warm up instance cache. On graceful restart it performs the following steps:
1. Restarts workers until the desirable amount of workers will be reached (configured with `--warmup-workers-ratio` option)
2. Waits until the instance cache becomes hot enough with respect to the old master's instance cache (configured with `--warmup-instance-cache-elements-ratio` option)
3. If it waits too long, continues workers restarting (configured with `--warmup-timeout` option)

Default behavior of graceful restart is unchanged.

Also here are 2 small improvements (last 2 commits):
- Don't send stats from new master when old master is still alive on restart. It prevents metrics distortion on long restarts.
- Increase max hanging worker time. Now it can't be less than script timeout and long script executions won't be treated 'hanging'.

### CODE REVIEW

File: server/php-master.cpp [hunk 1]
```diff
@@ -85,6 +86,8 @@ struct master_data_t {
   int ask_http_fd_generation;
   int sent_http_fd_generation;
 
+  uint64_t instance_cache_elements_stored;
```
[the remark this answers is not in the document]
Comment (DrDet) [reply server/php-master.cpp]: fixed

File: server/php-engine.cpp [hunk 2]
```diff
@@ -2659,6 +2704,9 @@ void parse_main_args(int argc, char *argv[]) {
   parse_option("profiler-log-prefix", required_argument, 2010, "set profier log path perfix");
   parse_option("mysql-db-name", required_argument, 2011, "database name of MySQL to connect");
   parse_option("net-dc-mask", required_argument, 2012, "a string formatted like '8=1.2.3.4/12' to detect a datacenter by ipv4");
+  parse_option("warmup-workers-part", required_argument, 2013, "configures master graceful restart: workers part (with respect to -f workers count) used for warm up on master graceful restart");
```
Comment (AlexK0) [inline server/php-engine.cpp:2707]: From my point of view, here we need short descriptions and clear options (for long descriptions we must use our documentation)
Please, consider the following
`warming-workers-ratio: the ratio of the instance cache warming workers during the graceful restart`
`warming-instance-case-elements-ratio: the ratio of the instance cache elements which makes the instance cache hot enough`
`warming-timeout: the maximum time for the instance cache warming up in seconds`

File: server/php-master-warmup.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  void set_warm_up_max_time(int warm_up_max_time) {
+    this->warm_up_max_time_ = std::chrono::seconds{warm_up_max_time};
+  }
+private:
+  double workers_part_for_warm_up_{1};
+  double target_instance_cache_elements_part_{0};
+  std::chrono::milliseconds warm_up_max_time_{5};
+
+  vk::SteadyTimer<std::chrono::milliseconds> timer{};
```
Comment (AlexK0) [inline server/php-master-warmup.h:49]: nit: please, be consistent, use `timer_` name as others above

File: server/php-master.cpp [hunk 4]
```diff
@@ -1897,15 +1696,21 @@ void run_master_on() {
...
     to_run = std::max(0, workers_n - total_workers);
 
-    if (other->valid_flag) {
-      int set_to_kill = std::max(std::min(MAX_KILL - other->dying_workers_n, other->running_workers_n), 0);
+    if (other->is_alive) {
+      auto &warm_up_ctx = WarmUpContext::get();
+      warm_up_ctx.try_start_warmup();
 
-      if (set_to_kill > 0) {
+      int set_to_kill = std::max(std::min(MAX_KILL - other->dying_workers_n, other->running_workers_n), 0);
```
Comment (AlexK0) [inline server/php-master.cpp:1706]: vk::clamp?

File: common/timer.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
+#pragma once
```
Comment (AlexK0) [inline common/timer.h:1]: licence header

File: server/php-master-restart.cpp [hunk 6]
```diff
@@ -0,0 +1,182 @@
+#include <cassert>
```
Comment (AlexK0) [inline server/php-master-restart.cpp:1]: licence header

File: server/php-master-restart.h [hunk 7]
```diff
@@ -0,0 +1,81 @@
+#pragma once
```
Comment (AlexK0) [inline server/php-master-restart.h:1]: licence header

File: server/php-master-warmup.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
+#pragma once
```
Comment (AlexK0) [inline server/php-master-warmup.h:1]: licence header

File: server/php-engine.cpp (hunk 2 above)
Comment (DrDet) [reply server/php-engine.cpp]: I agree, but from my point of view `--warmup-...` looks better than `--warming-...`. Fixed

File: server/php-engine.cpp [hunk 9]
```diff
@@ -2354,6 +2356,21 @@ void init_logname(const char *src) {
...
+    result = std::stod(optarg);
+  } catch (const std::exception &e) {
+    kprintf("%s option parse error: %s", option_name, e.what());
+    return result;
+  }
+  if (min_value <= result && result <= max_value) {
+    return result;
+  }
+  kprintf("%s should be in [%f; %f]", option_name, min_value, max_value);
+  return result;
```
Comment (AlexK0) [inline server/php-engine.cpp:2371]: return nan?

Comment (AlexK0) [inline server/php-engine.cpp:2371]: return nan?

Comment (DrDet) [reply server/php-engine.cpp]: fixed

**Review by AlexK0 — APPROVED** [review]

**Pull request merged**